gh-90949: add Expat API to prevent XML deadly allocations (#139234)
python/cpython: Doc/library/pyexpat.rst

XMLParser Objects
-----------------

> .. class:: xmlparser

Reviewer: Maybe as a side note:

The above does not match with user experience (and hence could be confusing):

```
In [1]: from xml.parsers.expat import XMLParserType

In [2]: XMLParserType.__name__
Out[2]: 'xmlparser'

In [3]: from xml.parsers.expat import xmlparser
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
Cell In[3], line 1
----> 1 from xml.parsers.expat import xmlparser

ImportError: cannot import name 'xmlparser' from 'xml.parsers.expat' (/usr/lib/python3.11/xml/parsers/expat.py)
```

So if that's on official class, maybe the import should be made to work. Or the docs made to use `XMLParserType` instead. (Or a new type `XMLParser` aliased to `XMLParserType` for backwards compatibility.)

I'm good with that being out of scope and for later.

Author: Mmmh. That's annoying. Actually `XMLParserType` is documented publicly as being the type being returned... So... I can instead document everything with respect to `XMLParserType` or see if I can use `xmlparser` as an alias to avoid too many amendments.

Author: Ok, it's too much for this PR. I'll just suppress the links for now and will think of it later. Probably, I'll expose `xmlparser` as an alias just for completeness, but let's address it separately.

Reviewer: @picnixz I understand :+1:

   .. versionadded:: 3.13

:class:`xmlparser` objects have the following methods to mitigate some
well-known XML vulnerabilities.

.. method:: xmlparser.SetAllocTrackerMaximumAmplification(max_factor, /)

   Sets the maximum amplification factor between direct input and bytes
   of dynamic memory allocated.

   By default, parsers objects have a maximum amplification factor of 100.

   The amplification factor is calculated as ``allocated / direct``
   while parsing, where ``direct`` is the number of bytes read from
   the primary document in parsing and ``allocated`` is the number
   of bytes of dynamic memory allocated in the parser hierarchy.

   The *max_factor* value must be a non-NaN :class:`float` value greater than
   or equal to 1.0. Amplifications factors greater than 100 can been observed
   near the start of parsing even with benign files in practice. As such, the

>    upper bound must be carefully chosen so to avoid false positives.

Reviewer: ```suggestion
   upper bound must be carefully chosen to avoid false positives.
```

>    An :exc:`ExpatError` is raised if this method is called by a non-root

Reviewer: ```suggestion
   An :exc:`ExpatError` is raised if this method is called for a non-root
```
…or "with" or "on"? The parser itself does not call it.

Author: Since it's a Python method, users would do `parser.<method>` so that's why I said "by". But happy to change to `on`.

Reviewer: @picnixz I'm not a native speaker of English but "by" feels too weird here to me. If we can agree to "on" I'm happy :+1:

   parser or if *max_factor* is outside the valid range. The corresponding

>    :attr:`~.ExpatError.lineno` and :attr:`~.ExpatError.column` should not be

Reviewer: `ExpatError` is class `xml.parsers.expat.ExpatError`. I'm having trouble understaing the "." in `~.ExpatError`but it may be just me. Would be curious to understand. Thanks!

Author: Actually, Sphinx complained that it wasn't able to find `ExpatError.lineno` attribute. The `.` tells Sphinx to search everywhere it can to find something that can be resolved as an attribute of the ExpatError class. I wanted to have a correct CI first, so I used that hammer, but I need to check either on Sphinx or here whether the `.. attribute::` directive has an issue (because, yes, the dot shouldn't have been necessary).

Reviewer: @picnixz thanks for the explanation :+1:

Author: Nevermind, it was me. I thought that both lineno and column were incorrect but only column was incorrect. The reason is... that it's offset and not column.

   used as they will have no special meaning.

   .. note::

      The maximum amplification factor is only considered if the threshold

>       specified by :meth:`.SetAllocTrackerActivationThreshold` is reached.

Reviewer: The current wording ignores the default threshold. The user does not have to adjust it, and it would still apply. How about:
```suggestion
      that can be adjusted by :meth:`.SetAllocTrackerActivationThreshold` is reached.
```

   .. versionadded:: 3.13

> :class:`xmlparser` objects have the following methods to mitigate some

Reviewer 2: `py:class reference target not found: xmlparser [ref.class]`

Seems that the `xmlparser` class is not documented, this introduces new reference warnings.

Author: Huh, why wasn't it caught by the CI? For now, I'll just suppress the link. We don't have a class description actually. We don't have a real description for this class, except some "XML Expat parser". What's important are the methods, not the constructor (which is actually not available; users need to create a parser through `expat.ParserCreate()`). I can add:

```rst
.. class:: xmlparser

   The type of an Expat XML parser created by :func:`ParserCreate`.
```

if you want.

common XML vulnerabilities.

> .. method:: xmlparser.SetAllocTrackerActivationThreshold(threshold, /)

Reviewer 2: For new methods, should we use snake_case naming?

Also, are there corresponding getters/properties to query the current values?

A

Author: > For new methods, should we use snake_case naming?

All methods follow the same naming as Expat so I prefer keeping the same logic. New methods still use the same convention. It's a bit like for unittest.TestCase where newly added methods are still named in camelCase (e.g., assertHasAttr, assertStartsWith)

> Also, are there corresponding getters/properties to query the current values?

No. It's also not exposed on Expat's side and the C structure is opaque.

   Sets the maximum amplification factor between direct input and bytes
   of dynamic memory allocated.

>    By default, parsers objects have a maximum amplification factor of 100.

Reviewer: ```suggestion
   By default, parser objects have a maximum amplification factor of 100.0.
```
(singular + for consistency?)

   The amplification factor is calculated as ``allocated / direct``
   while parsing, where ``direct`` is the number of bytes read from
   the primary document in parsing and ``allocated`` is the number
   of bytes of dynamic memory allocated in the parser hierarchy.

   The *max_factor* value must be a non-NaN :class:`float` value greater than

>    or equal to 1.0. Amplifications factors greater than 100 can been observed

Reviewer: ```suggestion
   or equal to 1.0. Amplifications factors greater than 100.0 can been observed
```
(for consistency?)

   near the start of parsing even with benign files in practice. As such, the
   activation threshold should be carefully chosen to avoid false positives.

   An :exc:`ExpatError` is raised if this method is called on a non-root
   parser or if *max_factor* is outside the valid range.
   The corresponding :attr:`~ExpatError.lineno` and :attr:`~ExpatError.offset`
   should not be used as they may have no special meaning.

   .. note::

      The maximum amplification factor is only considered if the threshold
      that can be adjusted :meth:`.SetAllocTrackerActivationThreshold` is
      exceeded.

   .. versionadded:: next

.. method:: xmlparser.SetAllocTrackerActivationThreshold(threshold, /)

   Sets the number of allocated bytes of dynamic memory needed to activate
   protection against disproportionate use of RAM.

>    By default, parsers objects have an allocation activation threshold of 64 MiB,

Reviewer: ```suggestion
   By default, parser objects have an allocation activation threshold of 64 MiB,
```
Singular

   .. versionadded:: 3.13

:class:`xmlparser` objects have the following methods to mitigate some
common XML vulnerabilities.

.. method:: xmlparser.SetAllocTrackerActivationThreshold(threshold, /)

   Sets the number of allocated bytes of dynamic memory needed to activate
   protection against disproportionate use of RAM.

   By default, parser objects have an allocation activation threshold of 64 MiB,
   or equivalently 67,108,864 bytes.

>    An :exc:`ExpatError` is raised if this method is called on a non-root parser.

Reviewer: I believe the term "root parser" may need to be explained or defined at least once, and I didn't see it defined anywhere

a)

```suggestion
   An :exc:`ExpatError` is raised if this method is called on a non-root parser,
   i.e. on a parser that has parents.
```

b)

```suggestion
   An :exc:`ExpatError` is raised if this method is called on a non-root parser,
   i.e. on a parser that has been created via :attr:`xmlparser.ExternalEntityParserCreate`.
```

(one more occurrence like this further down)

@picnixz what do you think?

Author: How about I actually define what a root parser is at one place so that I don't need to re-mention it everywhere?

Reviewer: @picnixz sure works for me :+1:

   The corresponding :attr:`~ExpatError.lineno` and :attr:`~ExpatError.offset`

>    should not be used as they may have no special meaning.

Reviewer: ```suggestion
   should not be used as they may have no particular meaning.
```
?

One more occurence like this further down

Author: We use `special meaning` 38 times in the docs while `particular meaning` only once so I'll keep my formulation. "special meaning" is still considered a synonym

Reviewer: @picnixz I have a hard time to agree on a language level, but we can have it your way :+1:
